### PR TITLE
TF2 porting: Conversion of feed forward attention reducer to TF2

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

Converted this function from TF1 to TF2 implementation
```
# # todo tf2: port all these functions to tf2
# def reduce_feed_forward_attention(current_inputs, hidden_size=256):
```

Converted function to `tf.keras.layers.Layer` subclass.

In addition to converting the function, added unit test `tests/ludwig/modules/test_attention.py` for the converted function.  Current scope of test is confirming correct output tensor shape for various input tensor sizes and specification of the reducer's `hidden_size` parameter.
```
 pytest -v /opt/project/tests/ludwig/modules/test_attention.py
=========================================================== test session starts ============================================================
platform linux -- Python 3.6.9, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /opt/project
plugins: xdist-2.0.0, forked-1.3.0, pycharm-0.7.0, typeguard-2.9.1
collected 24 items

../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-128-128] PASSED                              [  4%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-128-256] PASSED                              [  8%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-128-512] PASSED                              [ 12%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-256-128] PASSED                              [ 16%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-256-256] PASSED                              [ 20%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-10-256-512] PASSED                              [ 25%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-128-128] PASSED                              [ 29%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-128-256] PASSED                              [ 33%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-128-512] PASSED                              [ 37%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-256-128] PASSED                              [ 41%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-256-256] PASSED                              [ 45%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[16-20-256-512] PASSED                              [ 50%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-128-128] PASSED                              [ 54%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-128-256] PASSED                              [ 58%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-128-512] PASSED                              [ 62%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-256-128] PASSED                              [ 66%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-256-256] PASSED                              [ 70%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-10-256-512] PASSED                              [ 75%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-128-128] PASSED                              [ 79%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-128-256] PASSED                              [ 83%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-128-512] PASSED                              [ 87%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-256-128] PASSED                              [ 91%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-256-256] PASSED                              [ 95%]
../../tests/ludwig/modules/test_attention.py::test_feed_forward_attention_reducer[32-20-256-512] PASSED                              [100%]

============================================================ 24 passed in 0.55s ============================================================
```

### CODE REVIEW

**Comment by jimthompson5802** [comment]
@w4nderlust 
The last set of commits completes all the work I'm aware of to convert from the Mixin design to the `SequenceReducer` and `SequenceListReducer` design.

One item to note:  Since `BaseOutputFeature` supports two reduction parameters: `reduce_input` and `reduce_dependencies`, two instances of the `SequenceReducer` objects were created.  One with the `reduce_input` parameter setting and called `self.reduce_sequence_input` and the other with the `reduce_dependencies` parameter and called `self.reduce_sequence_dependencies`.

Assuming the `travis-ci` run comes out clean, this PR is ready for review and merge.

**Comment by jimthompson5802** [comment]
@w4nderlust The last set of commits implemented the following:
* Removed `SequenceListReducer` class and all associated code.
* Enhanced `test_dependencies.py` unit test to test all possible `reduce_dependencies` parameter settings: `['sum', 'mean', 'avg', 'max', 'concat', 'last']`

I believe this finally covers all the unit tests required for `SequenceReducer` and `FeedForwardAttentionReducer` refactoring.  

Assuming `travis-ci` run is clean, this is ready for review and merge.

File: ludwig/modules/reduction_modules.py [hunk 1]
```diff
@@ -73,34 +91,9 @@ def dont_reduce(sequence, **kwargs):
     'avg': reduce_mean,
     'max': reduce_max,
     'concat': reduce_concat,
-    'attention': reduce_feed_forward_attention,
+    'attention': FeedForwardAttentionReducer(),
+    'null': dont_reduce,  # todo tf2 do we need this value,added to support text_feature()
```
Comment (w4nderlust) [inline ludwig/modules/reduction_modules.py:95]: no we don't need it, let's modify to another valid value.

File: tests/integration_tests/test_dependencies.py [hunk 2]
```diff
@@ -42,8 +43,15 @@
         [BATCH_SIZE, SEQ_SIZE, HIDDEN_SIZE]
     ]
 )
-def test_multiple_dependencies(hidden_shape, dependent_hidden_shape,
-                               dependent_hidden_shape2):
+@pytest.mark.parametrize(
+    'reduce_dependencies', ['sum', 'mean', 'avg', 'max', 'concat', 'last']
```
Comment (w4nderlust) [inline tests/integration_tests/test_dependencies.py:47]: why not attention?

Comment (jimthompson5802) [reply tests/integration_tests/test_dependencies.py:47]: Only reason I had those values was what I read in the Ludwig documentation:
```
reduce_dependencies (default sum): defines how to reduce the output of a dependent feature that is not a vector, but a matrix or a higher order tensor, on the first dimension 9second if you count the batch dimension). Available values are: sum, mean or avg, max, concat (concatenates along the first dimension), last (returns the last vector of the first dimension).
```
I can add `attention`.  I'll add this to the running list I'm maintaining re: updating the ludwig docs with this new value

File: ludwig/modules/reduction_modules.py (hunk 1 above)
Comment (jimthompson5802) [reply ludwig/modules/reduction_modules.py:95]: OK...I'll modify `text_feature` to use one of the existing keys, e..g., `None`

File: tests/integration_tests/test_dependencies.py (hunk 2 above)
Comment (jimthompson5802) [reply tests/integration_tests/test_dependencies.py]: added 'attention` and resolved errors that were unique to the `attention` reducer

**Pull request merged**